lib/src/inline_parser.dart

```diff
@@ -572,146 +570,414 @@ class StrikethroughSyntax extends TagSyntax {
+      parser.advanceBy(-1);
+      return _addReferenceLink(parser, state, text);
+    }
+
+    if (char == $lbracket) {
+      parser.advanceBy(1);
+      // Maybe a reference link.
+      if (parser.pos + 1 < parser.source.length &&
+          parser.charAt(parser.pos + 1) == $rbracket) {
+        // Maybe a shortcut reference link.
+        parser.advanceBy(1);
+        return _addReferenceLink(parser, state, text);
+      }
+      var label = _parseReferenceLinkLabel(parser);
+      if (label != null) {
+        return _addReferenceLink(parser, state, label);
+      } else {
+        return false;
+      }
+    }
+
+    // The link text (inside `[...]`) was not followed with a opening `(` nor
+    // an opening `[`. Perhaps just a simple shortcut reference link (`[...]`).
+
+    // Back up back to the `]`.
+    return _addReferenceLink(parser, state, text);
   }
 
-  /// Given that [match] has matched both a title and URL, creates an `<a>`
-  /// [Element] for it.
-  Element _createElement(InlineParser parser, Match match, TagState state) {
-    var link = getLink(parser, match, state);
-    if (link == null) return null;
+  /// Resolve a possible reference link.
+  ///
+  /// Uses [linkReferences], [linkResolver], and [_createNode] to try to
+  /// resolve [label] and [state] into a [Node].
+  Node _resolveReferenceLink(
+      String label, TagState state, Map<String, LinkReference> linkReferences) {
+    var linkReference = linkReferences[label];
+    if (linkReference != null) {
+      return _createNode(state, linkReference.destination, linkReference.title);
+    } else {
+      // This link has no reference definition. But we allow users of the
+      // library to specify a custom resolver function ([linkResolver]) that
+      // may choose to handle this. Otherwise, it's just treated as plain
+      // text.
+      return linkResolver == null ? null : linkResolver(label);
+    }
+  }
 
+  /// Create the node represented by a Markdown link.
+  Node _createNode(TagState state, String destination, String title) {
     var element = new Element('a', state.children);
+    element.attributes['href'] = escapeAttribute(
+        destination.replaceAll(r'\(', '(').replaceAll(r'\)', ')'));
+    if (title != null && title.isNotEmpty) {
+      element.attributes['title'] = escapeAttribute(title);
+    }
+    return element;
+  }
 
-    element.attributes["href"] = escapeHtml(link.destination);
-    if (link.title != null) {
-      element.attributes['title'] = escapeHtml(link.title);
+  // Add a reference link node to [parser]'s AST.
+  //
+  // Returns whether the link was added successfully.
+  bool _addReferenceLink(InlineParser parser, TagState state, String label) {
+    var element =
+        _resolveReferenceLink(label, state, parser.document.linkReferences);
+    if (element == null) {
+      return false;
     }
+    parser.addNode(element);
+    parser.start = parser.pos;
+    _pendingStatesAreActive = false;
+    return true;
+  }
```

Comment:
I think I did more or less what you suggest here.

I've kept most logic in `LinkSyntax.onMatchEnd` though, as it is shared by `ImageSyntax` (meaning `ImageSyntax` doesn't override). `onMatchEnd` calls `_tryAddReferenceLink` in a few places, which _is_ overriden. In `LinkSyntax._tryAddReferenceLink`, if a reference link was parsed, then `_pendingStatesAreActive` is set to false, but not in `ImageSyntax._tryAddReferenceLink`. This is because, for some reason, ["an image description may contain links"](http://spec.commonmark.org/0.28/#images) (I assume only for non-HTML reasons).